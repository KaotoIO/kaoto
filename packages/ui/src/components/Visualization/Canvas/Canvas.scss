#topology-resize-panel {
  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v5-c-drawer__splitter::after {
    border: none;
  }
}
