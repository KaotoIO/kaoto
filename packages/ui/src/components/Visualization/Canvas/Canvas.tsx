import { Icon } from '@patternfly/react-core';
import { CatalogIcon } from '@patternfly/react-icons';
import {
  GRAPH_LAYOUT_END_EVENT,
  Model,
  SELECTION_EVENT,
  TopologyControlBar,
  TopologyControlButton,
  TopologyView,
  VisualizationProvider,
  VisualizationSurface,
  action,
  createTopologyControlButtons,
  defaultControlButtonsOptions,
} from '@patternfly/react-topology';
import {
  FunctionComponent,
  PropsWithChildren,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import layoutHorizontalIcon from '../../../assets/layout-horizontal.png';
import layoutVerticalIcon from '../../../assets/layout-vertical.png';
import { useLocalStorage } from '../../../hooks';
import { LocalStorageKeys } from '../../../models';
import { BaseVisualCamelEntity } from '../../../models/visualization/base-visual-entity';
import { CatalogModalContext } from '../../../providers/catalog-modal.provider';
import { VisibleFlowsContext } from '../../../providers/visible-flows.provider';
import { VisualizationEmptyState } from '../EmptyState';
import { CanvasSideBar } from './CanvasSideBar';
import { CanvasDefaults } from './canvas.defaults';
import { CanvasEdge, CanvasNode, LayoutType } from './canvas.models';
import { ControllerService } from './controller.service';
import { FlowService } from './flow.service';

interface CanvasProps {
  contextToolbar?: ReactNode;
  entities: BaseVisualCamelEntity[];
}

export const Canvas: FunctionComponent<PropsWithChildren<CanvasProps>> = ({ entities, contextToolbar }) => {
  /** State for @patternfly/react-topology */
  const [selectedIds, setSelectedIds] = useState<string[]>([]);
  const [selectedNode, setSelectedNode] = useState<CanvasNode | undefined>(undefined);
  const [nodes, setNodes] = useState<CanvasNode[]>([]);
  const [activeLayout, setActiveLayout] = useLocalStorage(LocalStorageKeys.CanvasLayout, CanvasDefaults.DEFAULT_LAYOUT);
  const [sidebarWidth, setSidebarWidth] = useLocalStorage(
    LocalStorageKeys.CanvasSidebarWidth,
    CanvasDefaults.DEFAULT_SIDEBAR_WIDTH,
  );

  /** Context to interact with the Canvas catalog */
  const catalogModalContext = useContext(CatalogModalContext);

  const controller = useMemo(() => ControllerService.createController(), []);
  const { visibleFlows } = useContext(VisibleFlowsContext)!;
  const shouldShowEmptyState = useMemo(() => {
    const areNoFlows = entities.length === 0;
    const areAllFlowsHidden = Object.values(visibleFlows).every((visible) => !visible);
    return areNoFlows || areAllFlowsHidden;
  }, [entities.length, visibleFlows]);

  const controlButtons = useMemo(() => {
    const customButtons: TopologyControlButton[] = catalogModalContext
      ? [
          {
            id: 'topology-control-bar-h_layout-button',
            icon: (
              <Icon>
                <img src={layoutHorizontalIcon} />
              </Icon>
            ),
            tooltip: 'Horizontal Layout',
            callback: action(() => {
              setActiveLayout(LayoutType.DagreHorizontal);
              controller.getGraph().setLayout(LayoutType.DagreHorizontal);
              controller.getGraph().reset();
              controller.getGraph().layout();
            }),
          },
          {
            id: 'topology-control-bar-v_layout-button',
            icon: (
              <Icon>
                <img src={layoutVerticalIcon} />
              </Icon>
            ),
            tooltip: 'Vertical Layout',
            callback: action(() => {
              setActiveLayout(LayoutType.DagreVertical);
              controller.getGraph().setLayout(LayoutType.DagreVertical);
              controller.getGraph().reset();
              controller.getGraph().layout();
            }),
          },
          {
            id: 'topology-control-bar-catalog-button',
            icon: <CatalogIcon />,
            tooltip: 'Open Catalog',
            callback: action(() => {
              catalogModalContext.setIsModalOpen(true);
            }),
          },
        ]
      : [];

    return createTopologyControlButtons({
      ...defaultControlButtonsOptions,
      zoomInCallback: action(() => {
        controller.getGraph().scaleBy(4 / 3);
      }),
      zoomOutCallback: action(() => {
        controller.getGraph().scaleBy(3 / 4);
      }),
      fitToScreenCallback: action(() => {
        controller.getGraph().fit(80);
      }),
      resetViewCallback: action(() => {
        controller.getGraph().reset();
        controller.getGraph().layout();
      }),
      legend: false,
      customButtons,
    });
  }, [catalogModalContext, controller, setActiveLayout]);

  const handleSelection = useCallback(
    (selectedIds: string[]) => {
      setSelectedIds(selectedIds);

      /** Current support for single selection at the moment */
      const selectedId = selectedIds[0];
      setSelectedNode(nodes.find((node) => node.id === selectedId));
    },
    [nodes],
  );

  /** Set up the controller one time */
  useEffect(() => {
    const localController = controller;
    const graphLayoutEndFn = action(() => {
      localController.getGraph().fit(80);
    });

    localController.addEventListener(SELECTION_EVENT, handleSelection);
    localController.addEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);

    return () => {
      localController.removeEventListener(SELECTION_EVENT, handleSelection);
      localController.removeEventListener(GRAPH_LAYOUT_END_EVENT, graphLayoutEndFn);
    };

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [handleSelection]);

  /** Draw graph */
  useEffect(() => {
    if (!Array.isArray(entities)) return;
    setSelectedNode(undefined);

    const nodes: CanvasNode[] = [];
    const edges: CanvasEdge[] = [];

    entities.forEach((entity) => {
      if (visibleFlows[entity.id]) {
        const { nodes: childNodes, edges: childEdges } = FlowService.getFlowDiagram(entity.toVizNode());
        nodes.push(...childNodes);
        edges.push(...childEdges);
      }
    });

    setNodes([...nodes]);

    const model: Model = {
      nodes,
      edges,
      graph: {
        id: 'g1',
        type: 'graph',
        layout: activeLayout,
      },
    };

    controller.fromModel(model, false);
  }, [activeLayout, controller, entities, visibleFlows]);

  useEffect(() => {
    const timeoutId = setTimeout(
      action(() => {
        controller.getGraph().fit(80);
      }),
      500,
    );

    return () => {
      clearTimeout(timeoutId);
    };
  }, [controller, selectedIds]);

  const handleCloseSideBar = useCallback(() => {
    setSelectedIds([]);
    setSelectedNode(undefined);
  }, []);

  const isSidebarOpen = useMemo(() => selectedNode !== undefined, [selectedNode]);

  return (
    <TopologyView
      defaultSideBarSize={sidebarWidth + 'px'}
      onSideBarResize={(width) => {
        setSidebarWidth(width);
      }}
      sideBarResizable
      sideBarOpen={isSidebarOpen}
      sideBar={<CanvasSideBar selectedNode={selectedNode} onClose={handleCloseSideBar} />}
      contextToolbar={contextToolbar}
      controlBar={<TopologyControlBar controlButtons={controlButtons} />}
    >
      <VisualizationProvider controller={controller}>
        {shouldShowEmptyState ? (
          <VisualizationEmptyState data-testid="visualization-empty-state" entitiesNumber={entities.length} />
        ) : (
          <VisualizationSurface state={{ selectedIds }} />
        )}
      </VisualizationProvider>
    </TopologyView>
  );
};
