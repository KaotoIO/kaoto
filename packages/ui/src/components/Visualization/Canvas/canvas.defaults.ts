import { NodeShape } from '@patternfly/react-topology';
import { LayoutType } from './canvas.models';

export class CanvasDefaults {
  static readonly DEFAULT_LAYOUT = LayoutType.DagreVertical;
  static readonly DEFAULT_NODE_SHAPE = NodeShape.rect;
  static readonly DEFAULT_NODE_DIAMETER = 75;
  static readonly DEFAULT_GROUP_PADDING = 50;
  static readonly DEFAULT_SIDEBAR_WIDTH = 500;
}
