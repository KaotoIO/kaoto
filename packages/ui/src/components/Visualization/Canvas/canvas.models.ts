import { EdgeModel, NodeModel } from '@patternfly/react-topology';
import { IVisualizationNode } from '../../../models/visualization/base-visual-entity';

export const enum LayoutType {
  BreadthFirst = 'BreadthFirst',
  Cola = 'Cola',
  ColaNoForce = 'ColaNoForce',
  Concentric = 'Concentric',
  Dagre = 'Dagre',
  Force = 'Force',
  Grid = 'Grid',
  ColaGroups = 'ColaGroups',
}

/**
 * The intention of these types is to isolate the usage of the
 * underlying rendering library tokens
 */
export interface CanvasNode extends NodeModel {
  parentNode?: string;
  data?: {
    vizNode?: IVisualizationNode;
  };
}

export interface CanvasEdge extends EdgeModel {
  source: string;
  target: string;
}

export type CanvasNodesAndEdges = { nodes: CanvasNode[]; edges: CanvasEdge[] };
