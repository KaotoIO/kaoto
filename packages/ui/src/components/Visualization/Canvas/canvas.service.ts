import {
  BreadthFirstLayout,
  ColaGroupsLayout,
  ColaLayout,
  ConcentricLayout,
  DagreLayout,
  DefaultEdge,
  DefaultGroup,
  DefaultNode,
  EdgeAnimationSpeed,
  EdgeStyle,
  ForceLayout,
  Graph,
  GraphComponent,
  GridLayout,
  Layout,
  Model,
  ModelKind,
  NodeShape,
  Visualization,
  withSelection,
} from '@patternfly/react-topology';
import { IVisualizationNode } from '../../../models/visualization/base-visual-entity';
import { CanvasEdge, CanvasNode, CanvasNodesAndEdges, LayoutType } from './canvas.models';

export class CanvasService {
  static readonly DEFAULT_LAYOUT = LayoutType.Dagre;
  static readonly DEFAULT_NODE_SHAPE = NodeShape.ellipse;
  static readonly DEFAULT_NODE_DIAMETER = 75;
  static readonly DEFAULT_GROUP_PADDING = 50;

  static nodes: CanvasNode[] = [];
  static edges: CanvasEdge[] = [];
  private static visitedNodes: string[] = [];

  static createController(): Visualization {
    const newController = new Visualization();

    newController.registerLayoutFactory(this.baselineLayoutFactory);
    newController.registerComponentFactory(this.baselineComponentFactory);

    const defaultModel: Model = {
      graph: {
        id: 'default',
        type: 'graph',
      },
    };

    newController.fromModel(defaultModel, false);

    return newController;
  }

  static baselineComponentFactory(kind: ModelKind, type: string) {
    switch (type) {
      case 'group':
        return DefaultGroup;
      default:
        switch (kind) {
          case ModelKind.graph:
            return GraphComponent;
          case ModelKind.node:
            return withSelection()(DefaultNode);
          case ModelKind.edge:
            return DefaultEdge;
          default:
            return undefined;
        }
    }
  }

  static baselineLayoutFactory(type: string, graph: Graph): Layout | undefined {
    switch (type) {
      case LayoutType.BreadthFirst:
        return new BreadthFirstLayout(graph);
      case LayoutType.Cola:
        return new ColaLayout(graph);
      case LayoutType.ColaNoForce:
        return new ColaLayout(graph, { layoutOnDrag: false });
      case LayoutType.Concentric:
        return new ConcentricLayout(graph);
      case LayoutType.Dagre:
        return new DagreLayout(graph, { rankdir: 'TB' });
      case LayoutType.Force:
        return new ForceLayout(graph);
      case LayoutType.Grid:
        return new GridLayout(graph);
      case LayoutType.ColaGroups:
        return new ColaGroupsLayout(graph, { layoutOnDrag: false });
      default:
        return new ColaLayout(graph, { layoutOnDrag: false });
    }
  }

  static getFlowDiagram(vizNode: IVisualizationNode): CanvasNodesAndEdges {
    this.nodes = [];
    this.edges = [];
    this.visitedNodes = [];
    this.appendNodesAndEdges(vizNode);

    return { nodes: this.nodes, edges: this.edges };
  }

  /** Method for iterating over all the IVisualizationNode and its children using a depth-first algorithm */
  private static appendNodesAndEdges(vizNodeParam: IVisualizationNode): void {
    if (this.visitedNodes.includes(vizNodeParam.id)) {
      return;
    }

    const node = this.getCanvasNode(vizNodeParam);

    /** Add node */
    this.nodes.push(node);
    this.visitedNodes.push(node.id);

    /** Add edges */
    this.edges.push(...this.getEdgesFromVizNode(vizNodeParam));

    /** Traverse the children nodes */
    const children = vizNodeParam.getChildren();
    if (children !== undefined) {
      children.forEach((child) => {
        this.appendNodesAndEdges(child);
      });
    }

    /** Traverse the next node */
    const nextNode = vizNodeParam.getNextNode();
    if (nextNode !== undefined) {
      this.appendNodesAndEdges(nextNode);
    }
  }

  private static getCanvasNode(vizNodeParam: IVisualizationNode): CanvasNode {
    /** Join the parent if exist to form a group */
    const parentNode =
      vizNodeParam.getParentNode()?.getChildren() !== undefined ? vizNodeParam.getParentNode()?.id : undefined;

    return this.getNode(vizNodeParam.id, {
      label: vizNodeParam.label,
      parentNode,
      data: { vizNode: vizNodeParam },
    });
  }

  private static getEdgesFromVizNode(vizNodeParam: IVisualizationNode): CanvasEdge[] {
    const edges: CanvasEdge[] = [];

    /** Connect to previous node if it doesn't have children */
    if (vizNodeParam.getPreviousNode() !== undefined && vizNodeParam.getPreviousNode()?.getChildren() === undefined) {
      edges.push(this.getEdge(vizNodeParam.getPreviousNode()!.id, vizNodeParam.id));
    }

    /** Connect to the parent if there is no previous node */
    if (vizNodeParam.getParentNode() !== undefined && vizNodeParam.getPreviousNode() === undefined) {
      edges.push(this.getEdge(vizNodeParam.getParentNode()!.id, vizNodeParam.id));
    }

    /** Connect to each leaf of the previous node */
    if (vizNodeParam.getPreviousNode() !== undefined && vizNodeParam.getPreviousNode()?.getChildren() !== undefined) {
      const leafNodesIds: string[] = [];
      vizNodeParam.getPreviousNode()!.populateLeafNodesIds(leafNodesIds);

      leafNodesIds.forEach((leafNodeId) => {
        edges.push(this.getEdge(leafNodeId, vizNodeParam.id));
      });
    }

    return edges;
  }

  private static getNode(
    id: string,
    options: { label?: string; parentNode?: string; data?: CanvasNode['data'] } = {},
  ): CanvasNode {
    return {
      id,
      type: 'node',
      label: options.label ?? id,
      parentNode: options.parentNode,
      data: options.data,
      width: this.DEFAULT_NODE_DIAMETER,
      height: this.DEFAULT_NODE_DIAMETER,
      shape: this.DEFAULT_NODE_SHAPE,
    };
  }

  private static getEdge(source: string, target: string): CanvasEdge {
    return {
      id: `${source}-to-${target}`,
      type: 'edge',
      source,
      target,
      edgeStyle: EdgeStyle.dashed,
      animationSpeed: EdgeAnimationSpeed.medium,
    };
  }
}
