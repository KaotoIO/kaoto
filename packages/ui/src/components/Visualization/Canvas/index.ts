export * from './Canvas';
export * from './CanvasSideBar';
export * from './DataFormatEditor';
export * from './StepExpressionEditor';
export * from './canvas.models';
