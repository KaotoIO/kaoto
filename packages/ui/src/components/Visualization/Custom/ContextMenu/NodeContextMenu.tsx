import { ArrowDownIcon, ArrowUpIcon, CodeBranchIcon, PlusIcon } from '@patternfly/react-icons';
import { ContextMenuSeparator, ElementModel, GraphElement } from '@patternfly/react-topology';
import { forwardRef, ReactElement } from 'react';
import { AddStepMode } from '../../../../models/visualization/base-visual-entity';
import { CanvasNode } from '../../Canvas/canvas.models';
import { ItemAddStep } from './ItemAddStep';
import { ItemDeleteGroup } from './ItemDeleteGroup';
import { ItemDeleteStep } from './ItemDeleteStep';
import { ItemDisableStep } from './ItemDisableStep';
import { ItemEnableAllSteps } from './ItemEnableAllSteps';
import { ItemInsertStep } from './ItemInsertStep';
import { ItemReplaceStep } from './ItemReplaceStep';

export const NodeContextMenuFn = (element: GraphElement<ElementModel, CanvasNode['data']>) => {
  const items: ReactElement[] = [];
  const vizNode = element.getData()?.vizNode;
  if (!vizNode) return items;

  const nodeInteractions = vizNode.getNodeInteraction();
  const childrenNodes = vizNode.getChildren();
  const isStepWithChildren = childrenNodes !== undefined && childrenNodes.length > 0;

  if (nodeInteractions.canHavePreviousStep) {
    items.push(
      <ItemAddStep
        key="context-menu-item-prepend"
        data-testid="context-menu-item-prepend"
        mode={AddStepMode.PrependStep}
        vizNode={vizNode}
      >
        <ArrowUpIcon /> Prepend
      </ItemAddStep>,
    );
  }
  if (nodeInteractions.canHaveNextStep) {
    items.push(
      <ItemAddStep
        key="context-menu-item-append"
        data-testid="context-menu-item-append"
        mode={AddStepMode.AppendStep}
        vizNode={vizNode}
      >
        <ArrowDownIcon /> Append
      </ItemAddStep>,
    );
  }
  if (nodeInteractions.canHavePreviousStep || nodeInteractions.canHaveNextStep) {
    items.push(<ContextMenuSeparator key="context-menu-separator-add" />);
  }

  if (nodeInteractions.canHaveChildren) {
    items.push(
      <ItemInsertStep
        key="context-menu-item-insert"
        data-testid="context-menu-item-insert"
        mode={AddStepMode.InsertChildStep}
        vizNode={vizNode}
      >
        <PlusIcon /> Add step
      </ItemInsertStep>,
    );
  }
  if (nodeInteractions.canHaveSpecialChildren) {
    items.push(
      <ItemInsertStep
        key="context-menu-item-insert-special"
        data-testid="context-menu-item-insert-special"
        mode={AddStepMode.InsertSpecialChildStep}
        vizNode={vizNode}
      >
        <CodeBranchIcon /> Add branch
      </ItemInsertStep>,
    );
  }
  if (nodeInteractions.canHaveChildren || nodeInteractions.canHaveSpecialChildren) {
    items.push(<ContextMenuSeparator key="context-menu-separator-insert" />);
  }

  if (nodeInteractions.canBeDisabled) {
    items.push(
      <ItemDisableStep key="context-menu-item-disable" data-testid="context-menu-item-disable" vizNode={vizNode} />,
    );
  }
  items.push(<ItemEnableAllSteps key="context-menu-item-enable-all" data-testid="context-menu-item-enable-all" />);

  if (nodeInteractions.canReplaceStep) {
    items.push(
      <ItemReplaceStep
        key="context-menu-item-replace"
        data-testid="context-menu-item-replace"
        vizNode={vizNode}
        loadActionConfirmationModal={isStepWithChildren}
      />,
    );
  }
  if (nodeInteractions.canBeDisabled || nodeInteractions.canReplaceStep) {
    items.push(<ContextMenuSeparator key="context-menu-separator-replace" />);
  }

  if (nodeInteractions.canRemoveStep) {
    items.push(
      <ItemDeleteStep key="context-menu-item-delete" data-testid="context-menu-item-delete" vizNode={vizNode} />,
    );
  }
  if (nodeInteractions.canRemoveFlow) {
    items.push(
      <ItemDeleteGroup
        key="context-menu-container-remove"
        data-testid="context-menu-item-container-remove"
        vizNode={vizNode}
      />,
    );
  }

  return items;
};

export const NodeContextMenu = forwardRef<HTMLDivElement, { element: GraphElement<ElementModel, CanvasNode['data']> }>(
  ({ element }, forwardedRef) => {
    return (
      <div data-testid="node-context-menu" ref={forwardedRef}>
        {NodeContextMenuFn(element)}
      </div>
    );
  },
);
