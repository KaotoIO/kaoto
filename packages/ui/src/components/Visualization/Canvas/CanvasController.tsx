import { VisualizationProvider } from '@patternfly/react-topology';
import { FunctionComponent, PropsWithChildren, ReactNode, useMemo } from 'react';
import { BaseVisualCamelEntity } from '../../../models/visualization/base-visual-entity';
import { Canvas } from './Canvas';
import './Canvas.scss';
import { ControllerService } from './controller.service';

interface CanvasProps {
  contextToolbar?: ReactNode;
  entities: BaseVisualCamelEntity[];
}

export const CanvasController: FunctionComponent<PropsWithChildren<CanvasProps>> = ({ entities }) => {
  const controller = useMemo(() => ControllerService.createController(), []);

  return (
    <VisualizationProvider controller={controller}>
      <Canvas entities={entities} />
    </VisualizationProvider>
  );
};
