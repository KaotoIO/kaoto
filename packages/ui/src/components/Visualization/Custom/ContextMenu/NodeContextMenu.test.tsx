import { ElementModel, GraphElement } from '@patternfly/react-topology';
import { render } from '@testing-library/react';
import { CanvasNode } from '../../Canvas';
import { NodeContextMenu } from './NodeContextMenu';
import { createVisualizationNode, IVisualizationNode, NodeInteraction } from '../../../../models';

describe('NodeContextMenu', () => {
  let element: GraphElement<ElementModel, CanvasNode['data']>;
  let vizNode: IVisualizationNode | undefined;
  let nodeInteractions: NodeInteraction;

  beforeEach(() => {
    nodeInteractions = {
      canHavePreviousStep: false,
      canHaveNextStep: false,
      canHaveChildren: false,
      canHaveSpecialChildren: false,
      canReplaceStep: false,
      canRemoveStep: false,
      canRemoveFlow: false,
      canBeDisabled: false,
    };
    vizNode = createVisualizationNode('test', {});
    jest.spyOn(vizNode, 'getNodeInteraction').mockReturnValue(nodeInteractions);
    element = {
      getData: () => {
        return { vizNode } as CanvasNode['data'];
      },
    } as unknown as GraphElement<ElementModel, CanvasNode['data']>;
  });

  it('should render an empty component when there is no vizNode', () => {
    vizNode = undefined;
    const { container } = render(<NodeContextMenu element={element} />);

    expect(container).toMatchSnapshot();
  });

  it('should render a PrependStep item if canHavePreviousStep is true', () => {
    nodeInteractions.canHavePreviousStep = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-prepend');

    expect(item).toBeInTheDocument();
  });

  it('should render an AppendStep item if canHaveNextStep is true', () => {
    nodeInteractions.canHaveNextStep = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-append');

    expect(item).toBeInTheDocument();
  });

  it('should render an InsertStep item if canHaveChildren is true', () => {
    nodeInteractions.canHaveChildren = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-insert');

    expect(item).toBeInTheDocument();
  });

  it('should render an InsertSpecialStep item if canHaveSpecialChildren is true', () => {
    nodeInteractions.canHaveSpecialChildren = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-insert-special');

    expect(item).toBeInTheDocument();
  });

  it('should render an ItemDisableStep item if canBeDisabled is true', () => {
    nodeInteractions.canBeDisabled = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-disable');

    expect(item).toBeInTheDocument();
  });

  it('should render an ItemReplaceStep item if canReplaceStep is true', () => {
    nodeInteractions.canReplaceStep = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-replace');

    expect(item).toBeInTheDocument();
  });

  it('should render an ItemDeleteStep item if canRemoveStep is true', () => {
    nodeInteractions.canRemoveStep = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-delete');

    expect(item).toBeInTheDocument();
  });

  it('should render an ItemDeleteGroup item if canRemoveFlow is true', () => {
    nodeInteractions.canRemoveFlow = true;
    const wrapper = render(<NodeContextMenu element={element} />);

    const item = wrapper.getByTestId('context-menu-item-container-remove');

    expect(item).toBeInTheDocument();
  });
});
