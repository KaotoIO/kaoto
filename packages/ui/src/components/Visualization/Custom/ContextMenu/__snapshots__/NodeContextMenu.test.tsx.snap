// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeContextMenu should render an empty component when there is no vizNode 1`] = `
<div>
  <div
    data-testid="node-context-menu"
  />
</div>
`;
